File: backend/src/main/java/org/currency_denomination/feature/domain/DenominationResult.java
package org.currency_denomination.feature.domain;

import java.util.Map;
import java.util.Optional;
import java.util.TreeMap;
import java.util.function.Function;

public class DenominationResult {
    private double valueForDenomination;
    private Optional<Double> valueForDifference;
    private Map<Double, Integer> denominations;
    private Map<Double, Integer> denominationsForDifference;
    private final CalculationType calculationType;
    private final Currency currency;

    public DenominationResult(
            double valueForDenomination,
            Function<Double, Map<Double, Integer>> calculator,
            Currency currency
    ) {
        this.valueForDenomination = valueForDenomination;
        this.valueForDifference = Optional.empty();
        this.denominations = calculator.apply(valueForDenomination);
        this.denominationsForDifference = new TreeMap<>();
        calculationType = CalculationType.DENOMINATION;
        this.currency = currency;
    }

    public DenominationResult(
            double valueForDenomination,
            double valueForDifference,
            Function<Double, Map<Double, Integer>> calculator,
            Currency currency
    ) {
        this.valueForDenomination = valueForDenomination;
        this.valueForDifference = Optional.of(valueForDifference);
        this.denominations = calculator.apply(valueForDenomination);
        this.denominationsForDifference = calculator.apply(valueForDifference);
        calculationType = CalculationType.DIFFERENCE;
        this.currency = currency;
    }

    public double getValueForDenomination() {
        return valueForDenomination;
    }

    public Optional<Double> getValueForDifference() {
        return valueForDifference;
    }

    public Map<Double, Integer> getDenominations() {
        return denominations;
    }

    public Map<Double, Integer> getDenominationsDifference() {
        return denominationsForDifference;
    }

    public CalculationType getCalculationType() {
        return calculationType;
    }

    public Currency getCurrency() {
        return currency;
    }
}

File: frontend/src/app/features/currencyDenomination/clientCalculationServiceService/DenominationResult.ts
import { CalculationType } from '../CalculationType';
import { CurrencySymbol } from '../CurrencySymbol';

export class DenominationResult {
  valueForDenomination: number;
  valueForDifference: number | null;
  denominations: Map<number, number>;
  denominationsForDifference: Map<number, number>;
  calculationType: CalculationType;
  currency: CurrencySymbol;

  private constructor(
    valueForDenomination: number,
    valueForDifference: number | null,
    denominations: Map<number, number>,
    denominationsForDifference: Map<number, number>,
    calculationType: CalculationType,
    currency: CurrencySymbol
  ) {
    this.valueForDenomination = valueForDenomination;
    this.valueForDifference = valueForDifference;
    this.denominations = denominations;
    this.denominationsForDifference = denominationsForDifference;
    this.calculationType = calculationType;
    this.currency = currency;
  }

  static fromSomethingForDenomination(
    valueForDenomination: number,
    calculationFunction: (value: number) => Map<number, number>,
    currencySymbol: CurrencySymbol
  ): DenominationResult {
    return new DenominationResult(
      valueForDenomination,
      null,
      calculationFunction(valueForDenomination),
      new Map<number, number>(),
      CalculationType.DENOMINATION,
      currencySymbol
    );
  }

  static fromSomethingWithDifference(
    valueForDenomination: number,
    valueForDifference: number,
    calculationFunction: (value: number) => Map<number, number>,
    currencySymbol: CurrencySymbol
  ): DenominationResult {
    return new DenominationResult(
      valueForDenomination,
      valueForDifference,
      calculationFunction(valueForDenomination),
      calculationFunction(valueForDifference),
      CalculationType.DIFFERENCE,
      currencySymbol
    );
  }
}
